# Retail Banking Risk & Fraud Analytics Platform

## Phase 2 : Data Cleaning & Data Quality Assessment

### Objective

The objective of this notebook is to evaluate the quality of the datasets before cleaning them.

Tasks:
- Load datasets
- Check duplicates
- Check missing values
- Analyze data types
- Memory usage
- Data quality assessment

Author: Md Aaris

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
transaction = pd.read_csv("../data/ieee_fraud/train_transaction.csv")
identity = pd.read_csv("../data/ieee_fraud/train_identity.csv")

In [3]:
print("=" * 60)
print("IEEE Transaction Dataset")
print("=" * 60)

print("Rows :", transaction.shape[0])
print("Columns :", transaction.shape[1])

print()

print("=" * 60)
print("IEEE Identity Dataset")
print("=" * 60)

print("Rows :", identity.shape[0])
print("Columns :", identity.shape[1])

IEEE Transaction Dataset
Rows : 590540
Columns : 394

IEEE Identity Dataset
Rows : 144233
Columns : 41


In [4]:
print("Duplicate Rows (Transaction):", transaction.duplicated().sum())
print("Duplicate Rows (Identity):", identity.duplicated().sum())

Duplicate Rows (Transaction): 0
Duplicate Rows (Identity): 0


In [5]:
missing_transaction = (
    transaction.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing_identity = (
    identity.isnull()
    .sum()
    .sort_values(ascending=False)
)

missing_transaction.head(20)

dist2    552913
D7       551623
D13      528588
D14      528353
D12      525823
D6       517353
D9       515614
D8       515614
V153     508595
V149     508595
V141     508595
V146     508595
V154     508595
V162     508595
V142     508595
V158     508595
V161     508595
V157     508595
V138     508595
V139     508595
dtype: int64

In [6]:
missing_identity.head(20)

id_24         139486
id_25         139101
id_07         139078
id_08         139078
id_21         139074
id_26         139070
id_23         139064
id_27         139064
id_22         139064
id_18          99120
id_04          77909
id_03          77909
id_33          70944
id_10          69307
id_09          69307
id_30          66668
id_32          66647
id_34          66428
id_14          64189
DeviceInfo     25567
dtype: int64

In [7]:
missing_percent = (
    transaction.isnull().mean() * 100
).sort_values(ascending=False)

missing_percent.head(20)

dist2    93.628374
D7       93.409930
D13      89.509263
D14      89.469469
D12      89.041047
D6       87.606767
D9       87.312290
D8       87.312290
V153     86.123717
V149     86.123717
V141     86.123717
V146     86.123717
V154     86.123717
V162     86.123717
V142     86.123717
V158     86.123717
V161     86.123717
V157     86.123717
V138     86.123717
V139     86.123717
dtype: float64

In [8]:
missing_percent_identity = (
    identity.isnull().mean() * 100
).sort_values(ascending=False)

missing_percent_identity.head(20)

id_24         96.708798
id_25         96.441868
id_07         96.425922
id_08         96.425922
id_21         96.423149
id_26         96.420375
id_23         96.416215
id_27         96.416215
id_22         96.416215
id_18         68.722137
id_04         54.016071
id_03         54.016071
id_33         49.187079
id_10         48.052110
id_09         48.052110
id_30         46.222432
id_32         46.207872
id_34         46.056034
id_14         44.503685
DeviceInfo    17.726179
dtype: float64

In [9]:
transaction.dtypes.value_counts()

float64    376
object      14
int64        4
Name: count, dtype: int64

In [10]:
identity.dtypes.value_counts()

float64    23
object     17
int64       1
Name: count, dtype: int64

In [11]:
transaction.memory_usage(deep=True).sum() / 1024**2

np.float64(2100.701410293579)

In [12]:
identity.memory_usage(deep=True).sum() / 1024**2

np.float64(157.6339931488037)

In [13]:
transaction.describe().T

,count,mean,std,min,25%,50%,75%,max
TransactionID,590540.0,3.282270e+06,1.704744e+05,2987000.000,3134634.750,3282269.500,3429904.25,3.577539e+06
isFraud,590540.0,3.499001e-02,1.837546e-01,0.000,0.000,0.000,0.00,1.000000e+00
TransactionDT,590540.0,7.372311e+06,4.617224e+06,86400.000,3027057.750,7306527.500,11246620.00,1.581113e+07
TransactionAmt,590540.0,1.350272e+02,2.391625e+02,0.251,43.321,68.769,125.00,3.193739e+04
card1,590540.0,9.898735e+03,4.901170e+03,1000.000,6019.000,9678.000,14184.00,1.839600e+04
...,...,...,...,...,...,...,...,...
V335,82351.0,5.916455e+01,3.876295e+02,0.000,0.000,0.000,0.00,5.512500e+04
V336,82351.0,2.853090e+01,2.745769e+02,0.000,0.000,0.000,0.00,5.512500e+04
V337,82351.0,5.535242e+01,6.684868e+02,0.000,0.000,0.000,0.00,1.040600e+05
V338,82351.0,1.511605e+02,1.095034e+03,0.000,0.000,0.000,0.00,1.040600e+05


# Initial Findings

## Transaction Dataset

- **Number of Rows:** 590,540
- **Number of Columns:** 394
- **Duplicate Rows:** 0
- **Missing Values:** A significant number of columns contain missing values, with several features having a high percentage of null values.
- **Memory Usage:** 2100.70 MB (~2.10 GB)

### Initial Observations

- The dataset contains **590,540 transaction records** and **394 features**, making it a high-dimensional dataset.
- No duplicate transaction records were found, indicating that each transaction is unique.
- Several engineered features, particularly the **V-series variables**, contain a large number of missing values and will require careful preprocessing.
- The dataset consists of both **numerical** and **categorical** features.
- The target variable **`isFraud`** is available, making this a supervised binary classification problem.
- Important features such as **Transaction Amount**, **Card Information**, **Email Domain**, **Address**, **Product Category**, and **Device-related variables** are expected to play a significant role in fraud detection.
- The dataset occupies approximately **2.10 GB of memory**, indicating that memory optimization will be important before model training.
- Overall, the transaction dataset is suitable for fraud detection after data cleaning, feature engineering, and preprocessing.

---

## Identity Dataset

- **Number of Rows:** 144,233
- **Number of Columns:** 41
- **Duplicate Rows:** 0
- **Missing Values:** Several identity-related features contain a high percentage of missing values.
- **Memory Usage:** 157.63 MB

### Initial Observations

- The dataset contains **144,233 customer identity records** with **41 features**.
- Each record is linked to the transaction dataset using the **TransactionID** column.
- No duplicate identity records were found.
- Identity-related features include device information, browser details, operating system, and engineered identity variables (`id_01` to `id_38`).
- Many identity attributes are missing because identity information is not available for every transaction.
- The identity dataset is relatively small (**157.63 MB**) compared to the transaction dataset and can be processed efficiently.
- Merging this dataset with the transaction dataset is expected to improve fraud detection performance.

---

## Next Steps

- Optimize memory usage.
- Analyze missing values in greater detail.
- Merge the Transaction and Identity datasets using **TransactionID**.
- Perform exploratory data analysis (EDA).
- Prepare the cleaned dataset for feature engineering and machine learning.

# Retail Banking Risk & Fraud Analytics Platform

# Phase 2: Data Preprocessing

## Objective

Prepare both datasets for Exploratory Data Analysis and Machine Learning by handling missing values, optimizing memory usage, and improving data quality.

Author: Md Aaris

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

In [2]:
transaction = pd.read_csv("../data/ieee_fraud/train_transaction.csv")
identity = pd.read_csv("../data/ieee_fraud/train_identity.csv")

In [3]:
print("Transaction Shape:", transaction.shape)
print("Identity Shape:", identity.shape)

Transaction Shape: (590540, 394)
Identity Shape: (144233, 41)


In [4]:
print("Transaction Memory:",
      round(transaction.memory_usage(deep=True).sum() / 1024**2, 2), "MB")

print("Identity Memory:",
      round(identity.memory_usage(deep=True).sum() / 1024**2, 2), "MB")

Transaction Memory: 2100.7 MB
Identity Memory: 157.63 MB


In [5]:
print("Transaction Duplicates:", transaction.duplicated().sum())
print("Identity Duplicates:", identity.duplicated().sum())

Transaction Duplicates: 0
Identity Duplicates: 0


## Missing Value Analysis

In [6]:
missing_transaction = (
    transaction.isnull()
    .mean()
    .mul(100)
    .sort_values(ascending=False)
)

missing_identity = (
    identity.isnull()
    .mean()
    .mul(100)
    .sort_values(ascending=False)
)

In [7]:
missing_transaction.head(20)

dist2    93.628374
D7       93.409930
D13      89.509263
D14      89.469469
D12      89.041047
D6       87.606767
D9       87.312290
D8       87.312290
V153     86.123717
V149     86.123717
V141     86.123717
V146     86.123717
V154     86.123717
V162     86.123717
V142     86.123717
V158     86.123717
V161     86.123717
V157     86.123717
V138     86.123717
V139     86.123717
dtype: float64

In [8]:
missing_identity.head(20)

id_24         96.708798
id_25         96.441868
id_07         96.425922
id_08         96.425922
id_21         96.423149
id_26         96.420375
id_23         96.416215
id_27         96.416215
id_22         96.416215
id_18         68.722137
id_04         54.016071
id_03         54.016071
id_33         49.187079
id_10         48.052110
id_09         48.052110
id_30         46.222432
id_32         46.207872
id_34         46.056034
id_14         44.503685
DeviceInfo    17.726179
dtype: float64

In [9]:
high_missing = missing_transaction[missing_transaction > 80]

medium_missing = missing_transaction[
    (missing_transaction > 30) &
    (missing_transaction <= 80)
]

low_missing = missing_transaction[
    missing_transaction <= 30
]

In [10]:
print("High Missing Columns :", len(high_missing))
print("Medium Missing Columns :", len(medium_missing))
print("Low Missing Columns :", len(low_missing))

High Missing Columns : 55
Medium Missing Columns : 137
Low Missing Columns : 202


In [11]:
high_missing_id = missing_identity[missing_identity > 80]

medium_missing_id = missing_identity[
    (missing_identity > 30) &
    (missing_identity <= 80)
]

low_missing_id = missing_identity[
    missing_identity <= 30
]

In [12]:
print("High Missing Columns :", len(high_missing_id))
print("Medium Missing Columns :", len(medium_missing_id))
print("Low Missing Columns :", len(low_missing_id))

High Missing Columns : 9
Medium Missing Columns : 10
Low Missing Columns : 22


## Data Type Analysis

In [13]:
print("Transaction Data Types")
transaction.dtypes.value_counts()

Transaction Data Types


float64    376
object      14
int64        4
Name: count, dtype: int64

In [14]:
print("Identity Data Types")
identity.dtypes.value_counts()

Identity Data Types


float64    23
object     17
int64       1
Name: count, dtype: int64

### Observations

- Numerical features dominate both datasets.
- Several categorical (object) features require encoding before model training.
- Data type optimization can significantly reduce memory usage.

## Memory Optimization

In [15]:
transaction_memory_before = transaction.memory_usage(deep=True).sum() / 1024**2
identity_memory_before = identity.memory_usage(deep=True).sum() / 1024**2

print(f"Transaction Memory Before: {transaction_memory_before:.2f} MB")
print(f"Identity Memory Before: {identity_memory_before:.2f} MB")

Transaction Memory Before: 2100.70 MB
Identity Memory Before: 157.63 MB


In [16]:
def reduce_memory_usage(df):

    for col in df.columns:

        col_type = df[col].dtype

        if col_type != object:

            c_min = df[col].min()
            c_max = df[col].max()

            if str(col_type)[:3] == "int":

                if c_min >= np.iinfo(np.int8).min and c_max <= np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)

                elif c_min >= np.iinfo(np.int16).min and c_max <= np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)

                elif c_min >= np.iinfo(np.int32).min and c_max <= np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)

            else:

                df[col] = df[col].astype(np.float32)

    return df

In [17]:
transaction = reduce_memory_usage(transaction)
identity = reduce_memory_usage(identity)

In [18]:
transaction_memory_after = transaction.memory_usage(deep=True).sum() / 1024**2
identity_memory_after = identity.memory_usage(deep=True).sum() / 1024**2

print(f"Transaction Memory After: {transaction_memory_after:.2f} MB")
print(f"Identity Memory After: {identity_memory_after:.2f} MB")

Transaction Memory After: 1241.85 MB
Identity Memory After: 144.43 MB


In [19]:
transaction_saved = (
    (transaction_memory_before - transaction_memory_after)
    / transaction_memory_before
) * 100

identity_saved = (
    (identity_memory_before - identity_memory_after)
    / identity_memory_before
) * 100

print(f"Transaction Memory Reduced: {transaction_saved:.2f}%")
print(f"Identity Memory Reduced: {identity_saved:.2f}%")

Transaction Memory Reduced: 40.88%
Identity Memory Reduced: 8.38%


### Memory Optimization Summary

- Successfully optimized numerical data types.
- Reduced dataset memory usage significantly.
- Optimized datasets will improve processing speed during analysis and model training.

## Missing Value Handling Strategy

### Transaction Dataset

- Columns with >80% missing values will be considered for removal.
- Columns with 30–80% missing values will be evaluated individually.
- Columns with <30% missing values will be imputed using appropriate techniques.

### Identity Dataset

- Very high missing columns will be dropped if they provide little information.
- Moderate missing columns will be imputed based on data type.
- Low missing columns will be filled using statistical methods or mode.

### Goal

Create a clean dataset while preserving as much useful information as possible for machine learning.

In [20]:
# Drop columns with more than 80% missing values

transaction = transaction.drop(columns=high_missing.index)

identity = identity.drop(columns=high_missing_id.index)

print("Transaction Shape:", transaction.shape)
print("Identity Shape:", identity.shape)

Transaction Shape: (590540, 339)
Identity Shape: (144233, 32)


In [21]:
print("Transaction Missing Values:")
print(transaction.isnull().sum().sort_values(ascending=False).head(10))

print("\nIdentity Missing Values:")
print(identity.isnull().sum().sort_values(ascending=False).head(10))

Transaction Missing Values:
V278    460110
V277    460110
V276    460110
V275    460110
V274    460110
V273    460110
V269    460110
V268    460110
V264    460110
V265    460110
dtype: int64

Identity Missing Values:
id_18    99120
id_03    77909
id_04    77909
id_33    70944
id_09    69307
id_10    69307
id_30    66668
id_32    66647
id_34    66428
id_14    64189
dtype: int64


In [22]:
# Fill numerical missing values with median

num_cols_transaction = transaction.select_dtypes(include=["int64", "float64"]).columns
num_cols_identity = identity.select_dtypes(include=["int64", "float64"]).columns

for col in num_cols_transaction:
    transaction[col] = transaction[col].fillna(transaction[col].median())

for col in num_cols_identity:
    identity[col] = identity[col].fillna(identity[col].median())

print("Numerical missing values filled successfully.")

Numerical missing values filled successfully.


In [23]:
# Fill categorical missing values with mode

cat_cols_transaction = transaction.select_dtypes(include=["object"]).columns
cat_cols_identity = identity.select_dtypes(include=["object"]).columns

for col in cat_cols_transaction:
    transaction[col] = transaction[col].fillna(transaction[col].mode()[0])

for col in cat_cols_identity:
    identity[col] = identity[col].fillna(identity[col].mode()[0])

print("Categorical missing values filled successfully.")

Categorical missing values filled successfully.


In [24]:
print("Transaction Missing Values Left:")
print(transaction.isnull().sum().sum())

print("Identity Missing Values Left:")
print(identity.isnull().sum().sum())

Transaction Missing Values Left:
64230006
Identity Missing Values Left:
577404


In [25]:
transaction.isnull().sum().sort_values(ascending=False).head(20)

V278    460110
V277    460110
V276    460110
V275    460110
V274    460110
V273    460110
V266    460110
V267    460110
V268    460110
V269    460110
V265    460110
V264    460110
V263    460110
V262    460110
V257    460110
V258    460110
V260    460110
V261    460110
V235    460110
V233    460110
dtype: int64

In [26]:
identity.isnull().sum().sort_values(ascending=False).head(20)

id_18            99120
id_03            77909
id_04            77909
id_09            69307
id_10            69307
id_32            66647
id_14            64189
id_13            16913
id_06             7368
id_05             7368
id_20             4972
id_19             4915
id_17             4864
id_02             3361
id_11             3255
TransactionID        0
id_16                0
id_15                0
id_12                0
id_01                0
dtype: int64

In [27]:
transaction.columns[transaction.isnull().any()]

Index(['card2', 'card3', 'card5', 'addr1', 'addr2', 'dist1', 'D1', 'D2', 'D3',
       'D4',
       ...
       'V312', 'V313', 'V314', 'V315', 'V316', 'V317', 'V318', 'V319', 'V320',
       'V321'],
      dtype='object', length=306)

In [28]:
identity.columns[identity.isnull().any()]

Index(['id_02', 'id_03', 'id_04', 'id_05', 'id_06', 'id_09', 'id_10', 'id_11',
       'id_13', 'id_14', 'id_17', 'id_18', 'id_19', 'id_20', 'id_32'],
      dtype='object')

## Conclusion

### Completed

- Loaded the IEEE-CIS Fraud Detection datasets.
- Explored dataset dimensions and structure.
- Analyzed missing values and categorized them based on severity.
- Checked duplicate records.
- Analyzed data types.
- Optimized memory usage by reducing numerical data types.
- Designed a missing-value handling strategy for future preprocessing.

### Key Findings

- The transaction dataset contains a large number of engineered features.
- Several V-series variables contain a high percentage of missing values.
- Numerical features dominate both datasets.
- Memory optimization reduced processing requirements significantly.

### Next Phase

In the next notebook, the datasets will undergo actual preprocessing, including:

- Handling missing values.
- Feature engineering.
- Encoding categorical variables.
- Preparing the datasets for Exploratory Data Analysis (EDA) and Machine Learning.

## Feature Classification

In [29]:
numerical_cols = transaction.select_dtypes(
    include=["int8","int16","int32","int64",
             "float16","float32","float64"]
).columns

categorical_cols = transaction.select_dtypes(
    include=["object"]
).columns

print("Numerical Columns :", len(numerical_cols))
print("Categorical Columns :", len(categorical_cols))

Numerical Columns : 325
Categorical Columns : 14


## Target Variable Analysis

In [30]:
transaction["isFraud"].value_counts()

isFraud
0    569877
1     20663
Name: count, dtype: int64

In [31]:
transaction["isFraud"].value_counts(normalize=True) * 100

isFraud
0    96.500999
1     3.499001
Name: proportion, dtype: float64

## Target Variable Observation

- The dataset is highly imbalanced.
- Approximately **96.5%** of the transactions are legitimate.
- Only **3.5%** of the transactions are fraudulent.
- During model development, class imbalance handling techniques such as class weights, undersampling, oversampling, or SMOTE may be considered.